contracts/mirror_lock/src/contract.rs

```diff
@@ -0,0 +1,279 @@
+            log(
+                "locked_amount",
+                position_locked_amount.to_string() + &config.base_denom,
+            ),
+            log("height", env.block.height),
+        ],
+        messages: vec![],
+        data: None,
+    })
+}
+
+pub fn unlock_position_funds<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    position_idx: Uint128,
+    check_lock_period: bool,
+) -> StdResult<HandleResponse> {
+    let config: Config = read_config(&deps.storage)?;
+    let sender_addr_raw: CanonicalAddr = deps.api.canonical_address(&env.message.sender)?;
+    let mut lock_info: PositionLockInfo = match read_position_lock_info(&deps.storage, position_idx)
+    {
+        Ok(lock_info) => lock_info,
+        Err(_) => {
+            return Err(StdError::generic_err(
+                "There are no locked funds for this position idx",
+            ))
+        }
+    };
+
+    // sparate the funds that can be unlocked
+    let (to_unlock, remaining): (Vec<(u64, Uint128)>, Vec<(u64, Uint128)>) = if check_lock_period {
+        // only position owner can unlock funds
+        if sender_addr_raw != lock_info.receiver {
+            return Err(StdError::unauthorized());
+        }
+        lock_info
+            .locked_funds
+            .iter()
+            .partition(|(lock_height, _)| env.block.height >= lock_height + config.lockup_period)
+    } else {
+        // only mint contract can force claim all funds, without checking lock period
+        if sender_addr_raw != config.mint_contract {
+            return Err(StdError::unauthorized());
+        }
+        (lock_info.locked_funds, vec![])
+    };
+
+    if to_unlock.is_empty() {
+        return Err(StdError::generic_err("Nothing to unlock"));
+    }
+
+    // calculate amount to unlock
+    let unlock_amount: u128 = to_unlock.iter().map(|item| item.1.u128()).sum();
+    let unlock_asset = Asset {
+        info: AssetInfo::NativeToken {
+            denom: config.base_denom.clone(),
+        },
+        amount: Uint128(unlock_amount),
+    };
+
+    if remaining.is_empty() {
+        remove_position_lock_info(&mut deps.storage, position_idx);
+    } else {
+        // update the lock info
+        lock_info.locked_funds = remaining;
+        store_position_lock_info(&mut deps.storage, &lock_info)?;
+    }
+
+    // decrease locked amount
+    total_locked_funds_store(&mut deps.storage).update(|current| {
+        let new_total = (current - Uint128(unlock_amount))?;
+        Ok(new_total)
+    })?;
+
+    let tax_amount: Uint128 = unlock_asset.compute_tax(&deps)?;
+
+    Ok(HandleResponse {
+        log: vec![
+            log("action", "unlock_shorting_funds"),
+            log("position_idx", position_idx.to_string()),
+            log("unlocked_amount", unlock_asset.to_string()),
+            log("tax_amount", tax_amount.to_string() + &config.base_denom),
+        ],
+        messages: vec![unlock_asset.clone().into_msg(
```

Comment:
maybe don't need to clone at final usage

Reply:
fixed :)